Upsell nudge: Redirect to signup destination if eligible

niranjan-uma-shankar opened the pull request:
#### Changes proposed in this Pull Request

* On post puchase upsell offer, clicking the Decline button should lead to the following:
    - Take to signup destination if eligible
    - Delete the signup destination cookie
* After opening an upsell offer using a direct link, like `/checkout/offer-quickstart-session`, clicking Decline should take to Customer Home

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Go through the signup flow at /start and purchase a Premium plan.
* On the post purchase upsell nudge, click on the "decline" button. You should be taken to Customer Home. The existing behaviour is that it would take to a "Thank You" page.
* On an existing account, open an upsell offer using the direct link, such as `/checkout/offer-quickstart-session`. Confirm that clicking the decline button will take root location `/`. 
* On an account having two sites, open `/checkout/offer-quickstart-session`. Clicking the decline button should take to root location `/`.
* On an existing account, upgrade to a higher plan. Verify that clicking the decline button on the post checkout upsell will take you to a Thank You page for the plan purchase.
* On an existing account, add an eCommerce plan to cart, exit checkout, and then navigate to `/checkout/offer-quickstart-session/{siteSlug}`. Clicking the decline button should take to root location `/` and no new cookies should be set. In production, there's a bug that with an eCommerce plan in cart, clicking decline will take to a thank-you page and then create a new `wpcom_signup_complete_destination` cookie.

niranjan-uma-shankar committed:
Check if destination is eligible for signup destination

niranjan-uma-shankar committed:
1. Clear signup destination cookie when declining the upsell offer 2. Direct load of upsell offer and subsequent decline should take to root path /

niranjan-uma-shankar committed:
Revert

sirbrillig commented on client/my-sites/checkout/composite-checkout/hooks/use-get-thank-you-url/get-thank-you-page-url.ts:
```diff
@@ -276,7 +276,7 @@ function getFallbackDestination( {
 	}
 
-	if ( siteSlug ) {
+	if ( siteSlug && ! isCartEmpty ) {
```
Thank you! Could you add a test to cover this change in `client/my-sites/checkout/composite-checkout/test/composite-checkout-thank-you.js`?

niranjan-uma-shankar replied:
Test case added, thanks.

niranjan-uma-shankar committed:
Update test case

niranjan-uma-shankar committed:
Redirect to root url '/' if cart and receipt ID is empty

niranjan-uma-shankar committed:
Fix the case of clicking Decline on the offer page when opening with a cart item

niranjan-uma-shankar committed:
Fix test cases

sirbrillig commented:
> On the post purchase upsell nudge, click on the "decline" button. You should be taken to Customer Home. The existing behaviour is that it would take to a "Thank You" page.

Works as described!

<img width="541" alt="Screen Shot 2021-02-02 at 12 06 48 PM" src="https://user-images.githubusercontent.com/2036909/106635882-55b9e500-654f-11eb-9eeb-6e30f56b3157.png">

> On an existing account, open an upsell offer using the direct link, such as /checkout/offer-quickstart-session. Confirm that clicking the decline button will take root location /.

I visited the url `/checkout/offer-quickstart-session` and clicked the decline button. Nothing happened and this error appeared in the console:

`Uncaught DOMException: Failed to execute 'replaceState' on 'History': A history state object with URL 'https://nopreviouspurchase/' cannot be created in a document with origin 'https://hash-cc70b62c7aa968e67916669f4e3dc5c195adf019.calypso.live' and URL 'https://hash-cc70b62c7aa968e67916669f4e3dc5c195adf019.calypso.live/checkout/offer-quickstart-session'.`

<img width="531" alt="Screen Shot 2021-02-02 at 12 07 54 PM" src="https://user-images.githubusercontent.com/2036909/106635971-7124f000-654f-11eb-9148-78e10d06881d.png">

I tried the url `/checkout/offer-quickstart-session/:my-site-here` and after clicking the decline button was redirected to customer home. This is the current behavior in production.

<img width="644" alt="Screen Shot 2021-02-02 at 12 10 58 PM" src="https://user-images.githubusercontent.com/2036909/106636242-bba66c80-654f-11eb-9081-1c377e3d2f17.png">

I'm not sure if either of those two are expected behavior. The error probably wouldn't happen in production if you can't reach the url `/checkout/offer-quickstart-session` (without a site slug) normally, **but it might be a good idea anyway to prevent it from returning the url `https://nopreviouspurchase/` URL in that case**. In production it redirects me to customer home.

> On an existing account, upgrade to a higher plan. Verify that clicking the decline button on the post checkout upsell will take you to a Thank You page for the plan purchase.

Works as described!

<img width="770" alt="Screen Shot 2021-02-02 at 12 14 50 PM" src="https://user-images.githubusercontent.com/2036909/106636958-44bda380-6550-11eb-94e6-a963dbfc914d.png">

niranjan-uma-shankar committed:
Fix merge error

niranjan-uma-shankar committed:
Return early if no previous purchase

niranjan-uma-shankar commented:
> I visited the url /checkout/offer-quickstart-session and clicked the decline button. Nothing happened and this error appeared in the console:

Thanks for catching that! I've added a fix for it in [09753ce](https://github.com/Automattic/wp-calypso/pull/49402/commits/09753cec1ce745e85457a25aee95cf79fdf96109). 

I also added a new test case in the PR description on a buggy behaviour I noticed in production (the last item `On an existing account, add an eCommerce plan to cart, exit checkout ... `).  The same commit [09753ce](https://github.com/Automattic/wp-calypso/pull/49402/commits/09753cec1ce745e85457a25aee95cf79fdf96109) will also fix this issue.

dzver commented:
I tried with a new user and site with the onboarding flow. Finishing the signup would lead me to /home and not to a thank-you page.

Adding a second site does not trigger the onboarding flow and this issue is not present.

Purchasing a plan from /plans redirects to the correct thank-you screen.

Adding an eCommerce plan to a site on a Premium plan when the user had more than one site resulted in the following warning:

<img width="1095" alt="Screenshot 2021-02-03 at 18 30 31" src="https://user-images.githubusercontent.com/82778/106777945-21791e00-664e-11eb-8cd0-05ecb7a08b21.png">

I don't see why the selectedSiteId would be required in this case given that `/checkout/concierge-session` already has site selection. I think this warning doesn't need to be addressed with this PR.

dzver commented:
Overall LGTM but the unit test that fails look related.

niranjan-uma-shankar committed:
Remove test case that is rendered outdated

niranjan-uma-shankar committed:
Fix incorrect variable

dzver reviewed (approved):
LGTM
